### PR TITLE
get source project id from merge request hardening

### LINKED ISSUE #164: Version 1.5.9 breaks gitlab notification for merge requests
With the plugin version 1.5.8 when a merge request was built in jenkins, the plugin would notify gitlab using the revision ref. Now gitlab could easily match the revision with any open merge request. That always worked for me.

Now with the plugin version 1.5.9, if I understand the modification correctly, it is trying to find the correct project id, I would assume in case the MR is from a fork. I can see in the code it parses the jenkins job name, split the string on character "-" and expect the last item of the split to be the merge request ID.

On some projects I have chosen the merge request build strategy "Both the current merge request revision and the merge request merged with the current target branch revision"

Which means for each merge request I get 2 jobs: MR-123-head and MR-123-merge.
Now with the blind string splitting it is trying to convert the strings head and merge to a number, fails to do so and it does not notify gitlab at all. At least it could revert to the previous strategy (revision ref) in that case.

```
java.lang.NumberFormatException: For input string: "head"
	at java.base/java.lang.NumberFormatException.forInputString(NumberFormatException.java:65)
	at java.base/java.lang.Integer.parseInt(Integer.java:652)
	at java.base/java.lang.Integer.parseInt(Integer.java:770)
	at io.jenkins.plugins.gitlabbranchsource.helpers.GitLabPipelineStatusNotifier.getSourceProjectId(GitLabPipelineStatusNotifier.java:213)
	at io.jenkins.plugins.gitlabbranchsource.helpers.GitLabPipelineStatusNotifier.sendNotifications(GitLabPipelineStatusNotifier.java:319)
	at io.jenkins.plugins.gitlabbranchsource.helpers.GitLabPipelineStatusNotifier.access$400(GitLabPipelineStatusNotifier.java:52)
	at io.jenkins.plugins.gitlabbranchsource.helpers.GitLabPipelineStatusNotifier$JobCheckOutListener.onCheckout(GitLabPipelineStatusNotifier.java:470)
```
```
java.lang.NumberFormatException: For input string: "merge"
	at java.base/java.lang.NumberFormatException.forInputString(NumberFormatException.java:65)
	at java.base/java.lang.Integer.parseInt(Integer.java:652)
	at java.base/java.lang.Integer.parseInt(Integer.java:770)
	at io.jenkins.plugins.gitlabbranchsource.helpers.GitLabPipelineStatusNotifier.getSourceProjectId(GitLabPipelineStatusNotifier.java:213)
	at io.jenkins.plugins.gitlabbranchsource.helpers.GitLabPipelineStatusNotifier.sendNotifications(GitLabPipelineStatusNotifier.java:319)
	at io.jenkins.plugins.gitlabbranchsource.helpers.GitLabPipelineStatusNotifier.access$400(GitLabPipelineStatusNotifier.java:52)
	at io.jenkins.plugins.gitlabbranchsource.helpers.GitLabPipelineStatusNotifier$JobCompletedListener.onCompleted(GitLabPipelineStatusNotifier.java:483)
```

### Reproduction steps

- Configure a project to use the Discover merge requests from origin strategy: Both the current merge request revision and the merge request merged with the current target branch revision
- Open a MR in gitlab, it will forever stay in the "Checking pipeline status" state

### PR DESCRIPTION
**Pull request by dcendents**
Fixes #164 

Now it will try all parts to convert the MR string from the project name into a number.

If the MR number cannot be parsed to a number or if there is an exception when fetching the MR from gitlab, then the project path is returned, which is always better than not notifying gitlab at all.

I've added unit tests to cover all the possible scenarions.
I've also tested it in jenkins and now it notifies gitlab correctly:

```
Oct 29, 2021 5:28:28 PM io.jenkins.plugins.gitlabbranchsource.helpers.GitLabPipelineStatusNotifier getSourceProjectId
INFO: Getting source project ID from MR
Oct 29, 2021 5:28:28 PM io.jenkins.plugins.gitlabbranchsource.helpers.GitLabPipelineStatusNotifier getSourceProjectId
INFO: Got source project ID from MR: 1504
```

image

### CODE REVIEW

**Comment by dcendents** [comment]
@jetersen Any chance you could review this PR and merge it. 1.5.9 has been broken for us and we run a snapshot version with this PR.

It would be useful to get it out of the way so I can work on changing the ids to be compatible with the new `gitlab4j-api`.

File: src/main/java/io/jenkins/plugins/gitlabbranchsource/helpers/GitLabPipelineStatusNotifier.java [hunk 1]
```diff
@@ -207,10 +207,21 @@ private static void logComment(Run<?, ?> build, TaskListener listener) {
...
         String[] jobFullNameParts = job.getFullName().split("-");
-        Integer mrId = Integer.parseInt(jobFullNameParts[jobFullNameParts.length - 1]);
+        Integer mrId = null;
+        for (int i = jobFullNameParts.length - 1; i >= 0; i--) {
+            try {
+                mrId = Integer.parseInt(jobFullNameParts[i]);
+                break;
+            } catch (NumberFormatException e) {
+            }
+        }
```
Comment (jetersen) [inline src/main/java/io/jenkins/plugins/gitlabbranchsource/helpers/GitLabPipelineStatusNotifier.java:220]: Why is this necessary? Split with limit should do the trick? unless there is something odd going odd here.

```java
        String[] jobFullNameParts = job.getFullName().split("-", 2);
```

Comment (dcendents) [reply src/main/java/io/jenkins/plugins/gitlabbranchsource/helpers/GitLabPipelineStatusNotifier.java:220]: I did not change that line...

Comment (jetersen) [reply src/main/java/io/jenkins/plugins/gitlabbranchsource/helpers/GitLabPipelineStatusNotifier.java:220]: Hmm not sure what to make of it than 😵

Comment (dcendents) [reply src/main/java/io/jenkins/plugins/gitlabbranchsource/helpers/GitLabPipelineStatusNotifier.java:220]: also check issue #164 for the details, my jobs are named `MR-123-head` and `MR-123-merge`, so yes I could assume if it is not the last part it will be the previous one, but I though it'd be safer to not assume (again) that project names will always follow that naming convention. As I understand that comes from another plugin so by being more tolerant it might work if something changes.

Comment (dcendents) [reply src/main/java/io/jenkins/plugins/gitlabbranchsource/helpers/GitLabPipelineStatusNotifier.java:220]: well version `1.5.9` introduced that logic where it gets the project ID from the MR id. Logic that did not consider people that choose the build strategy `Both the current merge request revision and the merge request merged with the current target branch revision`.

Now I'm trying to fix that by:
1. Parse the project name differently, not to assume the last part of the project name is the MR number
2. If for whatever reason we cannot get the project id, revert to the old logic (of `1.5.8`) of using the project name instead.

Comment (jetersen) [reply src/main/java/io/jenkins/plugins/gitlabbranchsource/helpers/GitLabPipelineStatusNotifier.java:220]: These names come from this plugin, so we know the format.

```java
String[] jobFullNameParts = job.getFullName().split("-", 3);
Integer mrId = jobFullNameParts.length == 3 ? jobFullNameParts[1] : jobFullNameParts[1];
```

Comment (dcendents) [reply src/main/java/io/jenkins/plugins/gitlabbranchsource/helpers/GitLabPipelineStatusNotifier.java:220]: This is the commit that broke it for us: https://github.com/jenkinsci/gitlab-branch-source-plugin/commit/cbe0d50c4b628bc669864403a651159ec56db8a7

**Comment by dcendents** [comment]
superseded by #187
